numba_cuda/numba/cuda/runtime/memsys.cuh

```diff
@@ -0,0 +1,17 @@
+#include <cuda/atomic>
+
+// Globally needed variables
+struct NRT_MemSys {
+    struct {
+      bool enabled;
+      cuda::atomic<size_t, cuda::thread_scope_device> alloc;
+      cuda::atomic<size_t, cuda::thread_scope_device> free;
+      cuda::atomic<size_t, cuda::thread_scope_device> mi_alloc;
+      cuda::atomic<size_t, cuda::thread_scope_device> mi_free;
+    } stats;
+  };
+
+/* The Memory System object */
+__device__ NRT_MemSys* TheMSys;
```

Comment:
I *think* this is initialized to 0 (and therefore the checks in nrt.cu like `if(TheMSys && ...)` are always safe. Am I thinking about that correctly?

Reply:
Also, if this is included into both memsys.cu and nrt.cu, is there a potential ODR violation here? Should this be `extern` in the header and defined in one of those two files?

Reply:
> I think this is initialized to 0 (and therefore the checks in nrt.cu like if(TheMSys && ...) are always safe. Am I thinking about that correctly?

Yes.

> Also, if this is included into both memsys.cu and nrt.cu, is there a potential ODR violation here? Should this be extern in the header and defined in one of those two files?

Currently `memsys.cu` and `nrt.cu` are compiled into two separate modules, so far ODR violation isn't an issue. This is a deliberate design decision to separate memsys stats method from NRT methods. I was hoping this can save us some compilation time when we turn on NRT, but do not turn on statistics. This design is debatable and could change.